File: apps/ios/Core/MeaningController.swift
import Foundation
import Observation

public struct MeaningRequest: Equatable, Sendable {
    public let sessionID: UUID
    public let passageID: String
    public let revisionKey: String
    public let text: String
    public let learningLanguageID: String
    public let meaningLanguage: String
    public init(sessionID: UUID, passage: Passage, learningLanguageID: String, meaningLanguage: String) {
        self.sessionID = sessionID; passageID = passage.id; revisionKey = passage.revisionKey
        text = passage.text; self.learningLanguageID = learningLanguageID; self.meaningLanguage = meaningLanguage
    }
    public var cacheKey: String { Self.cacheKey(revisionKey: revisionKey, language: meaningLanguage) }
    public static func cacheKey(revisionKey: String, language: String) -> String { "caption2/" + language + "::" + revisionKey }
    /// Caption text sent to the translation helper. Must match what the learner sees for this revision.
    public var translationInput: String { Self.translationInput(for: text) }
    public static func translationInput(for text: String) -> String { text }
    func sharesContext(with other: Self) -> Bool {
        sessionID == other.sessionID && passageID == other.passageID &&
        learningLanguageID == other.learningLanguageID && meaningLanguage == other.meaningLanguage
    }
}

public struct MeaningResult: Sendable {
    public let text: String
    public let inputTokens: Int
    public let outputTokens: Int
    public init(text: String, inputTokens: Int = 0, outputTokens: Int = 0) {
        self.text = text; self.inputTokens = inputTokens; self.outputTokens = outputTokens
    }
}

/// Keeps one translation in flight while coalescing growing transcript fragments.
@MainActor @Observable public final class MeaningController {
    public private(set) var text = ""
    public private(set) var isLoading = false
    public private(set) var error: String?
    @ObservationIgnored public var onResult: ((MeaningRequest, MeaningResult) -> Void)?
    @ObservationIgnored private let translate: @MainActor (MeaningRequest, @escaping @MainActor (String) -> Void) async throws -> MeaningResult
    @ObservationIgnored private let delay: Duration
    @ObservationIgnored private var desired: MeaningRequest?
    @ObservationIgnored private var rendered: MeaningRequest?
    @ObservationIgnored private var displayed: MeaningRequest?
    @ObservationIgnored private var lastDispatchedAt: ContinuousClock.Instant?
    @ObservationIgnored private var translationID: UUID?
    @ObservationIgnored private var worker: Task<Void, Never>?
    @ObservationIgnored private var generation = UUID()

    public init(delay: Duration = .milliseconds(450), translate: @escaping @MainActor (MeaningRequest) async throws -> MeaningResult) {
        self.delay = delay; self.translate = { request, _ in try await translate(request) }
    }
    public init(delay: Duration = .milliseconds(450), streaming: @escaping @MainActor (MeaningRequest, @escaping @MainActor (String) -> Void) async throws -> MeaningResult) {
        self.delay = delay; self.translate = streaming
    }
    deinit { worker?.cancel() }

    public func update(_ request: MeaningRequest, cached: String? = nil) {
        let changedContext = desired.map { !$0.sharesContext(with: request) } ?? true
        if changedContext { reset() }
        desired = request
        if let cached, !cached.isEmpty {
            cancelWorker(); text = cached; rendered = request; displayed = request; error = nil; return
        }
        if rendered == request { return }
        // Do not display a translation of text that was subsequently corrected.
        if let displayed, !request.text.hasPrefix(displayed.text) { text = ""; self.rendered = nil; self.displayed = nil }
        if worker == nil && error == nil { begin() }
    }
    public func reset() {
        cancelWorker(); desired = nil; rendered = nil; displayed = nil; lastDispatchedAt = nil; text = ""; error = nil
    }
    public func retry() {
        guard desired != nil else { return }
        cancelWorker(); error = nil; begin()
    }
    private func cancelWorker() {
        generation = UUID(); translationID = nil; worker?.cancel(); worker = nil; isLoading = false
    }
    private func begin() {
        guard desired != nil, worker == nil else { return }
        isLoading = true
        let token = generation
        worker = Task { [weak self] in
            guard let self else { return }
            do {
                let elapsed = self.lastDispatchedAt.map { $0.duration(to: .now) } ?? .zero
                try await Task.sleep(for: max(.zero, self.delay - elapsed))
                guard token == self.generation, !Task.isCancelled, let request = self.desired else { return }
                self.lastDispatchedAt = .now
                let translationID = UUID(); self.translationID = translationID
                // Retain the readable prefix until the next stream has caught up.
                let minimumPartialLength = self.text.count
                let result = try await self.translate(request) { [weak self] partial in
                    guard let self, token == self.generation, self.translationID == translationID, !Task.isCancelled,
                          let latest = self.desired, latest.sharesContext(with: request),
                          latest.text.hasPrefix(request.text), !partial.isEmpty, partial.count >= minimumPartialLength else { return }
                    self.text = partial; self.displayed = request
                }
                guard token == self.generation, !Task.isCancelled, let latest = self.desired else { return }
                self.translationID = nil
                guard !result.text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { throw MeaningError.empty }
                self.onResult?(request, result)
                if latest.sharesContext(with: request), latest.text.hasPrefix(request.text) {
                    self.text = result.text; self.rendered = request; self.displayed = request
                }
                self.worker = nil; self.isLoading = false
                if latest != request { self.begin() }
            } catch {
                guard token == self.generation, !Task.isCancelled else { return }
                self.translationID = nil; self.worker = nil; self.isLoading = false
                if self.rendered != self.desired { self.text = ""; self.displayed = nil }
                self.error = error.localizedDescription
            }
        }
    }
    private enum MeaningError: LocalizedError {
        case empty
        var errorDescription: String? { "The translation came back empty. Please try again." }
    }
}

File: apps/android/app/src/main/java/chat/mural/core/MeaningController.kt
package chat.mural.core

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.CoroutineStart
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

data class MeaningRequest(
    val sessionID: String, val passageID: String, val revisionKey: String, val text: String,
    val learningLanguageID: String, val meaningLanguage: String,
) {
    constructor(sessionID: String, passage: Passage, learningLanguageID: String, meaningLanguage: String) :
        this(sessionID, passage.id, passage.revisionKey, passage.text, learningLanguageID, meaningLanguage)

    val cacheKey get() = cacheKey(revisionKey, meaningLanguage)

    /** Caption text sent to the translation helper. Must match what the learner sees for this revision. */
    val translationInput get() = translationInput(text)

    fun sharesContext(other: MeaningRequest) = sessionID == other.sessionID && passageID == other.passageID &&
        learningLanguageID == other.learningLanguageID && meaningLanguage == other.meaningLanguage

    companion object {
        fun cacheKey(revisionKey: String, language: String) = "caption2/$language::$revisionKey"
        fun translationInput(text: String): String = text
    }
}

data class MeaningResult(val text: String, val inputTokens: Int = 0, val outputTokens: Int = 0)

class MeaningInputLimitException : Exception("This caption is too long to translate in one request.")

class EmptyMeaningException : Exception("The translation came back empty.")

/** Waits for a sentence or quiet transcript, then keeps one translation in flight. */
class MeaningController(
    private val scope: CoroutineScope,
    private val delayMillis: Long = 450,
    private val incompleteDelayMillis: Long = 1800,
    private val canRetryFailure: (Throwable) -> Boolean = { false },
    private val retryDelay: (Throwable) -> Long? = { null },
    private val minimumSpacingMillis: Long = 2500,
    private val now: () -> Long = { System.nanoTime() / 1_000_000 },
    private val stream: (suspend (MeaningRequest, (String) -> Unit) -> MeaningResult)? = null,
    private val translate: suspend (MeaningRequest) -> MeaningResult,
) {
    var text = ""; private set
    var isLoading = false; private set
    var error: Throwable? = null; private set
    var onResult: ((MeaningRequest, MeaningResult) -> Unit)? = null
    var onChange: (() -> Unit)? = null

    private var desired: MeaningRequest? = null
    private var rendered: MeaningRequest? = null
    private var displayed: MeaningRequest? = null
    private var activeTranslation: Long? = null
    private var nextTranslation = 0L
    private var worker: Job? = null
    private var generation = 0
    private var translating = false
    private var finalRequested = false
    private var finalRetryAvailable = false
    private var sessionEnded = false
    private var desiredUpdatedAt = 0L
    private var lastDispatchedAt: Long? = null
    private var automaticRetries = 0
    private var retryWindowStartedAt: Long? = null
    private var retryNotBefore = 0L
    private var pendingFailure: Throwable? = null

    fun update(request: MeaningRequest, cached: String? = null, utteranceComplete: Boolean = false, conversationEnded: Boolean = false) {
        if (desired?.sharesContext(request) != true) {
            val sameConversation = desired?.let { it.sessionID == request.sessionID &&
                it.learningLanguageID == request.learningLanguageID && it.meaningLanguage == request.meaningLanguage } == true
            if (translating && sameConversation) {
                // A new display passage waits for the current request instead of starting a second helper.
                rendered = null; displayed = null; text = ""; error = null; finalRequested = false; finalRetryAvailable = false; sessionEnded = false
                automaticRetries = 0; retryWindowStartedAt = null; retryNotBefore = 0; pendingFailure = null
            } else reset()
        }
        val changed = desired != request
        desired = request
        if (changed) desiredUpdatedAt = now()
        val newBoundary = (utteranceComplete && (!finalRequested || changed)) || (conversationEnded && !sessionEnded)
        if (newBoundary) finalRetryAvailable = true
        finalRequested = utteranceComplete || conversationEnded
        sessionEnded = sessionEnded || conversationEnded
        // Only a conclusive pre-provider rejection may be reopened automatically at a turn boundary.
        if (newBoundary && error?.let(canRetryFailure) == true) {
            val failure = error!!
            prepareAutomaticRetry(failure, retryDelay(failure) ?: 0)
            finalRetryAvailable = false
        }
        if (!cached.isNullOrEmpty()) {
            if (!translating) cancelWorker()
            text = cached; rendered = request; displayed = request; error = null; isLoading = false
            pendingFailure = null; retryNotBefore = 0
            onChange?.invoke(); return
        }
        if (rendered == request) return
        displayed?.let { if (!request.text.startsWith(it.text)) { text = ""; rendered = null; displayed = null } }
        // A waiting timer follows the latest fragment; an admitted request is never cancelled by speech.
        if (!translating && (changed || utteranceComplete || conversationEnded)) cancelWorker()
        if (worker == null && error == null) begin()
        onChange?.invoke()
    }

    fun reset() {
        cancelWorker(); desired = null; rendered = null; displayed = null; text = ""; error = null
        finalRequested = false; finalRetryAvailable = false; sessionEnded = false
        automaticRetries = 0; retryWindowStartedAt = null; retryNotBefore = 0; pendingFailure = null
        onChange?.invoke()
    }

    fun retry() {
        if (desired == null) return
        if (translating) return
        cancelWorker(); error = null; finalRequested = true; finalRetryAvailable = false
        automaticRetries = 0; retryWindowStartedAt = null; retryNotBefore = 0; pendingFailure = null
        begin()
        onChange?.invoke()
    }

    private fun cancelWorker() {
        generation++; activeTranslation = null; worker?.cancel(); worker = null; isLoading = false; translating = false
    }

    private fun endsSentence(value: String): Boolean =
        value.trimEnd().trimEnd('"', '\'', '”', '’', '»', ')').lastOrNull() in setOf('.', '!', '?', '。', '！', '？', '…')

    private fun prepareAutomaticRetry(failure: Throwable, wait: Long): Boolean {
        val current = now()
        val started = retryWindowStartedAt ?: current
        if (wait !in 0..30_000 || automaticRetries >= 3 || current - started + wait >= 30_000) return false
        retryWindowStartedAt = started; automaticRetries++
        retryNotBefore = current + wait; pendingFailure = failure; error = null
        return true
    }

    private fun begin() {
        if (desired == null || rendered == desired || worker != null) return
        isLoading = true
        val token = generation
        worker = scope.launch(start = CoroutineStart.LAZY) {
            var dispatched: MeaningRequest? = null
            try {
                val initial = desired ?: return@launch
                val current = now()
                val quietUntil = desiredUpdatedAt + if (finalRequested) 0 else if (endsSentence(initial.text)) delayMillis else incompleteDelayMillis
                val pacedUntil = lastDispatchedAt?.let { it + minimumSpacingMillis } ?: current
                delay((maxOf(quietUntil, pacedUntil, retryNotBefore) - current).coerceAtLeast(0))
                val request = desired
                if (token != generation || request == null) return@launch
                pendingFailure?.let { failure ->
                    if (now() - (retryWindowStartedAt ?: now()) >= 30_000) throw failure
                }
                pendingFailure = null; retryNotBefore = 0
                translating = true; lastDispatchedAt = now(); dispatched = request
                val translation = ++nextTranslation; activeTranslation = translation
                // Retain the readable prefix until the next stream has caught up.
                val minimumPartialLength = text.length
                val result = if (stream == null) translate(request) else stream.invoke(request) { partial ->
                    val latest = desired
                    if (token == generation && activeTranslation == translation && latest != null && rendered != latest &&
                        latest.sharesContext(request) && latest.text.startsWith(request.text) && partial.isNotEmpty() && partial.length >= minimumPartialLength) {
                        text = partial; displayed = request; onChange?.invoke()
                    }
                }
                val latest = desired
                if (token != generation || latest == null) return@launch
                activeTranslation = null
                if (result.text.isBlank()) throw EmptyMeaningException()
                onResult?.invoke(request, result)
                if (rendered != latest && latest.sharesContext(request) && latest.text.startsWith(request.text)) {
                    text = result.text; rendered = request; displayed = request
                }
                worker = null; isLoading = false; translating = false
                if (latest != request) begin()
                onChange?.invoke()
            } catch (e: CancellationException) {
                if (token == generation) throw e
            } catch (e: Exception) {
                if (token != generation) return@launch
                activeTranslation = null; worker = null; isLoading = false; translating = false
                if (rendered == desired) { error = null; onChange?.invoke(); return@launch }
                if (dispatched != null && desired?.sharesContext(dispatched) == false) {
                    // Do not retry or render the old passage; continue with the separately requested new one.
                    begin(); onChange?.invoke(); return@launch
                }
                error = e
                // A prefix translation is not the meaning of a completed, longer utterance.
                if (rendered != desired) text = ""
                // End may arrive while a previously admitted request is completing or being rejected.
                val wait = retryDelay(e) ?: if (finalRequested && finalRetryAvailable && canRetryFailure(e)) 0L else null
                finalRetryAvailable = false
                if (wait != null && prepareAutomaticRetry(e, wait)) begin()
                onChange?.invoke()
            }
        }
        worker?.start()
    }
}
